core/commands/store/upload/ds/sessions.go

```diff
@@ -0,0 +1,251 @@
+package ds
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/TRON-US/go-btfs/core"
+	"github.com/TRON-US/go-btfs/core/commands/storage"
+	"github.com/TRON-US/go-btfs/core/escrow"
+	sessionpb "github.com/TRON-US/go-btfs/protos/session"
+
+	config "github.com/TRON-US/go-btfs-config"
+	iface "github.com/TRON-US/interface-go-btfs-core"
+	escrowpb "github.com/tron-us/go-btfs-common/protos/escrow"
+	"github.com/tron-us/protobuf/proto"
+
+	"github.com/google/uuid"
+	"github.com/ipfs/go-datastore"
+	"github.com/looplab/fsm"
+	cmap "github.com/orcaman/concurrent-map"
+	"github.com/prometheus/common/log"
+)
+
+const (
+	sessionKeyPrefix   = "/btfs/%s/v0.0.1/renter/sessions/%s/"
+	sessionInMemKey    = sessionKeyPrefix
+	sessionMetadataKey = sessionKeyPrefix + "metadata"
+	session_status_key = sessionKeyPrefix + "status"
+)
+
+var (
+	sessionsInMem = cmap.New()
+	fsmEvents     = fsm.Events{
+		{Name: "e-submit", Src: []string{"init"}, Dst: "submit"},
+		{Name: "e-pay", Src: []string{"submit"}, Dst: "pay"},
+		{Name: "e-guard", Src: []string{"pay"}, Dst: "guard"},
+		{Name: "e-wait-upload", Src: []string{"guard"}, Dst: "wait-upload"},
+		{Name: "e-complete", Src: []string{"wait-upload"}, Dst: "complete"},
+		{Name: "e-error", Src: []string{"init", "submit", "pay", "guard", "wait-upload"}, Dst: "error"},
+	}
+)
+
+type Session struct {
+	Id        string
+	PeerId    string
+	Datastore datastore.Datastore
+	Context   context.Context
+	cancel    context.CancelFunc
+	Config    *config.Config
+	N         *core.IpfsNode
+	Api       iface.CoreAPI
+	fsm       *fsm.FSM
+}
+
+type SessionInitParams struct {
+	Context     context.Context
+	Config      *config.Config
+	N           *core.IpfsNode
+	Api         iface.CoreAPI
+	Datastore   datastore.Datastore
+	RenterId    string
+	FileHash    string
+	ShardHashes []string
+}
+
+func GetSession(sessionId string, peerId string, params *SessionInitParams) (*Session, error) {
+	// NewSession
+	if sessionId == "" {
+		sessionId = uuid.New().String()
+	}
+	k := fmt.Sprintf(sessionInMemKey, peerId, sessionId)
+	var s *Session
+	if tmp, ok := sessionsInMem.Get(k); ok {
+		s = tmp.(*Session)
+		status, err := s.GetStatus()
+		if err != nil {
+			return nil, err
+		}
+		s.fsm.SetState(status.Status)
```

Comment:
I am entirely sure this is goroutine-safe, this is a get function with a lot of SetXXX - maybe some thing next steps will need to handle

Ideally, in my opinion, these settings should be part of `s.GetStatus` or in a single call...